src/python/pants/backend/native/tasks/link_shared_libraries.py

```diff
@@ -154,21 +141,30 @@ def _make_link_request(self,
     deps = self._retrieve_single_product_at_target_base(native_target_deps_product, vt.target)
 
     all_compiled_object_files = []
-
     for dep_tgt in deps:
-      self.context.log.debug("dep_tgt: {}".format(dep_tgt))
-      object_files = self._retrieve_single_product_at_target_base(compiled_objects_product, dep_tgt)
-      self.context.log.debug("object_files: {}".format(object_files))
-      object_file_paths = object_files.file_paths()
-      self.context.log.debug("object_file_paths: {}".format(object_file_paths))
-      all_compiled_object_files.extend(object_file_paths)
-
-    return LinkSharedLibraryRequest.with_external_libs_product(
+      if compiled_objects_product.get(dep_tgt):
```

Comment:
Unsure whether we should accept the lack of a product for that target and not error out -- I think we should be able to assume the product for the compiled output exists in the mapping (even if the product contains no files), unless there's some polymorphism happening here I'm missing.

Reply:
Ok, that actually makes sense. Maybe a comment on the fact that `dep_tgt` could be either an external library or a source library could be useful, as that spooky action happens at a distance in `NativeCompile` and isn't immediately evident in this file.